File: facultate/sd/T3/radixsort/radixsort.py
from math import log


def radix(inp, BASE=1000):
    if len(inp) == 0:
        return []
    digits = (int(log(max(inp), BASE)) + 1)

    def setup():
        d = {}
        for i in range(BASE):
            d[i] = []
        return d

    def dicToList(d):
        l = []
        for values in d.values():
            for value in values:
                l.append(value)
        return l

    key = 1  # adica cifra unitatilor (abcd // key = d)
    for _ in range(digits):
        d = setup()
        for i in inp:
            d[(i//key) % BASE].append(i)
        inp = dicToList(d)
        key *= BASE
    return inp


if __name__ == "__main__":
    with open('radixsort.in', 'r') as f:
        inp = [int(x.strip()) for x in f.readline().split()]
        n = inp[0]
        a = inp[1]
        b = inp[2]
        c = inp[3]
    arr = [0] * n
    arr[0] = b
    for i in range(1, n):
        arr[i] = (a * arr[i-1] + b) % c

    # arr = radix(arr, int(log(c, 2)))
    arr.sort()

    with open('radixsort.out', 'w') as f:
        for x in arr[0::10]:
            f.write(str(x) + ' ')

File: facultate/sd/T3/radixsort/radixsort.cpp
// https://www.infoarena.ro/problema/radixsort
// 30p

#include <bits/stdc++.h>
using namespace std;

template <typename T>
ostream &operator<<(ostream &os, const vector<T> &v) {
    for (int i = 0; i < v.size(); ++i) {
        os << v[i];
        if (i != v.size() - 1) os << ", ";
    }
    return os;
}

map<int, vector<int>> setup(int base) {
    map<int, vector<int>> d;
    for (int i = 0; i < base; ++i)
        d[i] = vector<int>();
    return d;
}

vector<int> mapToVector(map<int, vector<int>> d) {
    vector<int> v;
    for (map<int, vector<int>>::iterator it = d.begin(); it != d.end(); ++it)
        for (auto e : it->second) v.push_back(e);
    return v;
}

vector<int> radix(vector<int> inp, int base = 100) {
    if (inp.size() == 0) return inp;
    int digits = (int)(floor(log(*max_element(inp.begin(), inp.end())) / log(base))) + 1;
    int key = 1;
    for (int _ = 0; _ < digits; ++_) {
        map<int, vector<int>> d = setup(base);
        for (auto i : inp)
            d[((int)i / key) % base].push_back(i);
        inp = mapToVector(d);
        key *= base;
    }
    return inp;
}

int main() {
    ifstream fin("radixsort.in", ios::in);
    ofstream fout("radixsort.out", ios::out);

    int n, a, b, c;
    fin >> n >> a >> b >> c;
    vector<int> arr(n);
    arr[0] = b;
    for (int i = 1; i < n; ++i)
        arr[i] = (a * arr[i - 1] + b) % c;

    arr = radix(arr);

    for (int i = 0; i < arr.size(); i += 10) {
        fout << arr[i];
        if (i != arr.size() - 1) fout << " ";
    }
}